### PR TITLE
Unify weight_scale_2 between gate_proj/up_proj (and w1/w3) in the HF export path for MOE models

### PR DESCRIPTION
**Pull request by Edwardf0t1**
### What does this PR do?
Unify `weight_scale_2` between `gate_proj/up_proj` (and `w1/w3`) in the HF export path for MOE models. Serving engines fuse these projections into a single `gate_up_proj` and require a shared scale; this takes the element-wise max of the two independent scales as a conservative choice that avoids overflow.

Type of change: ? Bug fix

### Usage

```python
# Add a code snippet demonstrating how to use this
```

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Automatic synchronization of quantization scaling between Mixture-of-Experts gate and up projections during model export for non‑fused MoE setups (e.g., Qwen MoE, DeepSeek).
* **Bug Fixes / Improvements**
  * Export now emits a brief notification when gate/up scaling values are adjusted to ensure consistent quantization.

### CODE REVIEW

File: modelopt/torch/export/quant_utils.py [hunk 1]
```diff
@@ -1154,6 +1154,35 @@ def postprocess_state_dict(
     return post_state_dict
 
 
+def max_gate_up_scales(tensors: dict[str, torch.Tensor]) -> int:
```
Comment (cjluo-nv) [inline modelopt/torch/export/quant_utils.py]: I think this may not work. We also need to requantize ws1 if the ws2 changes. I think we need to do it here instead: https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/export/unified_export_hf.py#L230

Comment (Edwardf0t1) [reply modelopt/torch/export/quant_utils.py]: I agree it's safer to sync the _amax on the weight quantizers of gate/up pairs. In @lukealonso's implementation it's done in both amax level and the post-processing stage, and he verified the accuracy which seems good.

**Comment by Edwardf0t1** [comment]
> https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/export/unified_export_hf.py#L396-L418 this logic is used to align w1 and w3 scales.
> 
> We might want to debug why this fails for qwen3.5

It failed for kimi-k2.5 as well. That's exactly why we needed this PR to fix it.

File: modelopt/torch/export/unified_export_hf.py [hunk 2]
```diff
@@ -791,6 +792,10 @@ def _export_transformers_checkpoint(
         quantized_state_dict, kv_cache_max_bound, kv_cache_format, is_modelopt_qlora
     )
 
+    tied = max_gate_up_scales(quantized_state_dict)
```
Comment (meenchen) [inline modelopt/torch/export/unified_export_hf.py]: Should we do this before compressing the weights, i.e., syncing the amax of gate and up? Is my understanding correct that the block-wise scales are still based on the separate global scales?

Comment (Edwardf0t1) [reply modelopt/torch/export/unified_export_hf.py]: You are right - updated to sync at the amax level.

**Comment by cjluo-nv** [comment]
If the goal for this is to serve a safe guard, can we do a check (like assert) so when the bug reproduces, we can see it immediately?

**Comment by Edwardf0t1** [comment]
> If the goal for this is to serve a safe guard, can we do a check (like assert) so when the bug reproduces, we can see it immediately?

Added a warning.

**Review by cjluo-nv — COMMENTED** [review]
## Follow-up: Correcting my earlier suggestion

I checked and `build_stacked_experts` (where the `weights_scaling_factor_2` assertion lives) is only called in the **TRT-LLM export path**, never from `unified_export_hf.py`. So my suggestion to put the fix there wouldn't help the HF export path — the gap is real.

That said, the core feedback about duplication still stands. Rather than a standalone function with its own traversal and `_GATE_UP_PAIRS` constant, consider:

1. **Reuse `get_expert_linear_names`** to derive the gate/up pair instead of hardcoding `_GATE_UP_PAIRS` — this keeps expert naming centralized.
2. **Place the sync inline** right before `_process_quantized_modules(model, ...)` is called (around line 778), possibly as a small loop using `is_moe` + `get_expert_linear_names` + the existing `get_experts_list` or `expert_getter` helpers. This avoids introducing a new traversal pattern.

The fix logic itself (element-wise max of amaxes) is fine — it's the packaging that could be tighter.

**Comment by Edwardf0t1** [comment]
> ## Concrete proposal
> Sorry for the confusion — here's exactly what I'd suggest:
> 
> ### What to change
> **Delete** the new `sync_moe_gate_up_amax` function and `_GATE_UP_PAIRS` constant from `layer_utils.py`. Instead, add the sync logic **inline** in `unified_export_hf.py` inside the existing expert loop (lines 699-750), which already calls `get_expert_linear_names` and iterates over experts. This avoids any new traversal or naming constants.
> 
> ### Where to put it
> In `_export_transformers_checkpoint`, right after the existing expert input-quantizer loop (after line 750), add something like:
> 
> ```python
>     # Sync gate/up weight quantizer amaxes for unfused MoE experts.
>     # Serving engines fuse gate+up into one projection and need a shared weight_scale_2.
>     synced = 0
>     for _, sub_module in model.named_modules():
>         if not (is_moe(sub_module) and hasattr(sub_module, "experts")):
>             continue
>         if not isinstance(sub_module.experts, collections.abc.Iterable):
>             continue
>         linear_names = get_expert_linear_names(sub_module)
>         # Only unfused 3-linear experts (e.g. gate_proj/down_proj/up_proj or w1/w2/w3)
>         if len(linear_names) \!= 3:
>             continue
>         gate_name, up_name = linear_names[0], linear_names[2]
>         for expert in sub_module.experts:
>             gate_wq = getattr(getattr(expert, gate_name, None), "weight_quantizer", None)
>             up_wq = getattr(getattr(expert, up_name, None), "weight_quantizer", None)
>             if gate_wq is None or up_wq is None:
>                 continue
>             gate_amax = getattr(gate_wq, "amax", None)
>             up_amax = getattr(up_wq, "amax", None)
>             if gate_amax is None or up_amax is None:
>                 continue
>             if not torch.equal(gate_amax, up_amax):
>                 shared = torch.max(gate_amax, up_amax)
>                 gate_wq.amax = shared
>                 up_wq.amax = shared.clone()
>                 synced += 1
>     if synced:
>         warnings.warn(
>             f"Synced {synced} MoE expert gate/up weight quantizer amax pair(s) "
>             f"by taking element-wise max for serving-engine fusion."
>         )
> ```
> 
> ### Why this is better
> * **No new function or constant in `layer_utils.py`** — zero diff there
> * **Uses `get_expert_linear_names`** to derive gate/up names — single source of truth
> * **Same traversal pattern** as the existing loop above it (lines 699-750) — consistent, easy to follow
> * **Same fix logic** you already wrote — just relocated
> 
> That's it — same behavior, less code, no duplication.

I think the suggestion has a real problem. Here's why:
The current ordering in _export_transformers_checkpoint is:

```
Step 1 (lines 699-750):  Handle uncalibrated expert input quantizers
Step 2 (line 755):        requantize_resmooth_fused_llm_layers(model)
                            └─ preprocess_linear_fusion → syncs gate/up weight amaxes
                               (but may miss experts not activated by dummy forward)
Step 3 (line 782):        sync_moe_gate_up_amax(model)  ← safety net
Step 4 (line 792):        _process_quantized_modules     ← quantizes weights using amaxes
```


If we inline the sync into Step 1 (lines 699-750), it runs before requantize_resmooth_fused_llm_layers in Step 2. That's a problem because:

1. preprocess_linear_fusion (called by Step 2) can resmooth weights and recalibrate amaxes via _update_pre_quant_scale, which resets amaxes and re-runs finish_stats_collection. So any sync done in Step 1 could be undone by Step 2.

2. The current placement as a safety net after Step 2 is deliberate — it catches exactly the pairs that requantize_resmooth_fused_llm_layers missed, without interfering with what it successfully handled.

3. The existing loop at 699-750 is specifically for input quantizer amaxes of uncalibrated experts. Mixing in weight quantizer amax sync conflates two independent concerns and would make the already-complex loop harder to reason about.

The one valid point in the suggestion is avoiding a redundant model traversal and the _GATE_UP_PAIRS constant. But the traversal cost is negligible (it's iterating named_modules(), not doing any computation), and _GATE_UP_PAIRS is a 2-line constant that maps cleanly to what get_expert_linear_names returns for unfused architectures.

My recommendation: keep sync_moe_gate_up_amax as a separate function in its current position (after requantize_resmooth_fused_llm_layers). The ordering is correct, the separation of concerns is clean, and the cost is trivial.
